doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -0,0 +1,145 @@
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+   
+.. include:: ../../common.defs
+
+.. highlight:: cpp
+.. default-domain:: cpp
+.. |MemArena| replace:: :class:`MemArena`
+
+.. _MemArena:
+
+MemArena
+*************
+
+|MemArena| provides a memory arena or pool for allocating memory. The intended use is for allocating many small chunks of memory - few, large allocations are best handled independently. The purpose is to amortize the cost of allocation of each chunk across larger allocations in a heap style. In addition the allocated memory is presumed to have similar lifetimes so that all of the memory in the arena can be de-allocatred en masse. This is a memory allocation style used by many cotainers - elements are small and allocated frequently, but all elements are discarded when the container itself is destroyed.
+
+Description
++++++++++++
+
+|MemArena| manages an internal list of memory blocks, out of which it provides allocated
+blocks of memory. When an instance is destructed all the internal blocks are also freed. The
+expected use of this class is as an embedded memory manager for a container class.
+
+To support coalescence and compaction of memory, the methods :func:`MemArena::freeze` and
+:func:`MemArena::thaw` are provided. These create in effect generations of memory allocation.
+Calling :func:`MemArena::freeze` marks a generation. After this call any further allocations will
+be in new internal memory blocks. The corresponding call to :func:`MemArena::thaw` cause older
+generations of internal memory to be freed. The general logic for the container would be to freeze,
+re-allocate and copy the container elements, then thaw. This would result in compacted memory
+allocation in a single internal block. The uses cases would be either a process static data
+structure after initialization (coalescing for locality performence) or a container that naturally
+re-allocates (such as a hash table during a bucket expansion). A container could also provide its
+own API for its clients to cause a coalesence.
+
+Other than freeze / thaw, this class does not offer any mechanism to release memory beyond its destruction. This is not an issue for either process globals or transient arenas.
+
+Internals
++++++++++
+
+|MemArena| opperates in *generations*. Each generation marks a seperate internal block of memory. From a generation, memory can be allocated in the form of a :class:`MemSpan`. Once a generation is exhausted all avaliable memory, or is explicitly "frozen" via :func:`MemArena::freeze`, all allocations are frozen and a new generation must be created. Allocations made in previous generations can still be used when frozen but it is only possible to allocate new memory chunks from the current generation. Thus, every generation that is not the current generation is implicitly frozen. 
```

Comment:
This isn't correct. A generation can have any number of memory blocks. Generations are only created by explicit API calls. Every time a new block is needed it is twice the size of the previous block in the generation.

Reply:
Yea. I misunderstood what `MemArena` was supposed to do. I thought it was the case that each block was a new generation. It should be that you can allocate new internal blocks, but are in the same generation. Generations should only be marked with explicit calls to `freeze`. I updated the PR with new allocation logic.